ocsmesh/utils.py

```diff
@@ -151,47 +152,167 @@ def geom_to_multipolygon(mesh):
     return MultiPolygon(polygon_collection)
 
 
-def get_boundary_segments(mesh):
-
+def get_boundary_vertices(mesh):
     coords = mesh.vert2['coord']
     boundary_edges = get_boundary_edges(mesh)
     boundary_verts = np.unique(boundary_edges)
     boundary_coords = coords[boundary_verts]
 
     vert_map = {
-            orig: new for new, orig in enumerate(boundary_verts)}
+        orig: new for new, orig in enumerate(boundary_verts)}
     new_boundary_edges = np.array(
         [vert_map[v] for v in boundary_edges.ravel()]).reshape(
-                boundary_edges.shape)
+        boundary_edges.shape)
 
     graph = sparse.lil_matrix(
-            (len(boundary_verts), len(boundary_verts)))
+        (len(boundary_verts), len(boundary_verts)))
     for vert1, vert2 in new_boundary_edges:
         graph[vert1, vert2] = 1
 
     n_components, labels = sparse.csgraph.connected_components(
-            graph, directed=False)
+        graph, directed=False)
 
-    segments = []
+    vertices = []
     for i in range(n_components):
         conn_mask = np.any(np.isin(
-                new_boundary_edges, np.nonzero(labels == i)),
-                axis=1)
+            new_boundary_edges, np.nonzero(labels == i)),
+            axis=1)
         conn_edges = new_boundary_edges[conn_mask]
-        this_segment = linemerge(boundary_coords[conn_edges].tolist())
-        if not this_segment.is_simple:
+        line = linemerge(boundary_coords[conn_edges].tolist())
+        bverts = boundary_verts[conn_edges].tolist()
+
+        if not line.is_simple:
             # Pinched nodes also result in non-simple linestring,
             # but they can be handled gracefully, here we are looking
             # for other issues like folded elements
-            test_polys = list(polygonize(this_segment))
-            if not test_polys:
+            poly = list(polygonize(line))
+            if not poly:
+                # The antartic can intersect itself because the most southern point is not a beginning or start.
+                line = wrap_artic_around_minimum_latitude(line)
+                poly = polygonize(line)
+            if not poly:
+                poly = unwrap_linestring_to_valid_polygon(line)
+
+            if not poly:
                 raise ValueError(
                     "Mesh boundary crosses itself! Folded element(s)!")
-        segments.append(this_segment)
+        vertices.append(bverts)
+
+    return vertices
 
+
+def get_boundary_segments(mesh, boundary_vertices=None):
+    coords = mesh.vert2['coord']
+    if boundary_vertices is None:
+        boundary_vertices = get_boundary_vertices(mesh)
+    segments = []
+    for segment_vertices in boundary_vertices:
+        segment = linemerge(coords[segment_vertices, :].tolist())
+        segments.append(segment)
     return segments
 
 
+def get_boundary_data(mesh):
```

Comment:
@SanderHulst can you please explain if and how does this `get_boundary_data`  function differ from the existing functionality in `mesh/mesh.py`?